keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -0,0 +1,343 @@
+  overload scenario today there is no assurance of priority for these
+  self-maintenance requests.
+
+- Priority Inversions.  In the course of serving request A, there are
+  some other requests B spawned --- directly or indirectly.  One
+  example is requests B that arrive over a loopback connection ---
+  such as requests issued by an admission plugin (e.g., ResourceQuota)
+  or any client-ish code in a registry strategy.  Another example is
+  requests issued by an external server that itself is serving
+  call-outs from an apiserver (e.g., admission web-hooks).  Other
+  examples include requests from an aggregated apiserver to create
+  TokenReview and SubjectAccessReview objects.  Today it is possible
+  that the very load imposed by request A crowds out requests B
+  involved in serving A.
+
+- Garbage Collector crowded out.  The garbage collector should keep up
+  with the workload, but in an overload situation today this is not
+  assured to happen.
+
+- Deployment of Doom.  We had a situation where a bug in the
+  Deployment controller caused it to run amuck under certain
+  circumstances, issuing requests in a tight loop.  We would like
+  controller bugs to not take the whole system down.
+
+- Kubelet Amuck.  The controller that runs amuck might not be a
+  central singleton, it could be a kubelet, kube-proxy, or other
+  per-node or otherwise multiplied controller.  In such a situtation
+  we would like only the guilty individual to suffer, not all its
+  peers and the rest of the system.
+
+- Overbearing or buggy tenants.  In a multi-tenant scenario, we would
+  like to prevent some tenants from crowding out the others.
+
+### Non-Goals
+
+This will be our first cut at a significant area of functionality, and
+our goals are deliberately modest.  We want to implement something
+useful but limited and get some experience before going further.  Our
+precise modesty has not been fully agreed.  Following is an initial
+stake in the ground.
+
+- No coordination between apiservers nor with a load balancer is
+  attempted.  In this KEP each apiserver independently protects
+  itself.  We imagine that later developments may add support for
+  informing load balancers about the load state of the apiservers.
+
+- The fairness does not have to be highly precise.  Any rough fairness
+  will be good enough.
+
+- WATCH and CONNECT requests are out of scope.  These are of a fairly
+  different nature than the others, and their management will be more
+  complex.  Also they are arguably less of an observed problem.
+
+- We are only concerned with protection of the CPU and memory of the
+  apiserver.  We are not concerned with etcd performance, nor output
+  network bandwidth, nor the ability of clients to consume output.
+
+- This KEP will not attempt auto-tuning the capacity limit(s).  Instead
+  the administrator will configure each apiserver's capacity limit(s),
+  analogously to how the max-in-flight limits are configured today.
```

Comment:
I think it is a goal that autotuning be possible/natural in the future, though--we can't start down a path that doesn't lead to autotuning.

Reply:
I think nothing in the current thinking rules out or is inimical to autotuning.  Autotuning would be a future enhancement that dynamically adjusts the resource limit(s) according to some appropriate considerations.